### DISCUSSION TITLE
Construction of record instance from on ascendant instance using non-destructive mutation

### DISCUSSION THREAD

**Discussion by zorgoz** [dotnet/csharplang#5872] [Language Ideas]
Currently non-destructive mutation can't take advantage of inheritance between records. But this could be really convenient.

Let's take for example following record types:
```
public record Person
{
    public string FullName { get; set; }
}

public record Student : Person
{
    public string Grade { get; set; }
}
```
or
```
public record Person(string FullName);

public record Student(string FullName, string Grade): Person(FullName);

```
Currently, the way to create a Student instance from Person is this:
```
var person = new Person { FullName = "XYZ" };
var student = new Student {FullName = person.FullName, Grade = "9"};
```
or
```
var person = new Person("XYZ");
var student = new Student(person.FullName, "9");
```
Meaning, that one needs to write assignments for all the properties that are present anyway in both types - except of course, for the ones that will have some new value. This is a cumbersome and error-prone. Non-destructive mutations allow only cloning of an instance of the same type. 

The syntax proposed by me is: **`var student = new Student from person with {Grade = "9"};`**
Semantics: a new descendant instance is created and all properties of the ascendant are cloned in the same way as the mutation does, except for the value overrides present in the `with` clause for properties can be declared in the descendant, or inherited from the ascendant. 

Of course these would be also valid:
```
var student = new Student from person;
var student = new Student from person with {};
var student = new Student from person with {FullName = "Buch Johns", Grade = "10"};
```

**Comment by HaloFour** [comment]
What would this compile down into?  What if `person` is already an instance of `Student`?

The `with` keyword doesn't make the new instance directly, it asks the object to clone itself and then it writes the updated properties:

```cs
var person2 = person1 with { FullName = "Bob" };

// compiles into
var person2 = person1.Clone();
person2.FullName = "Bob";
```

**Reply by zorgoz** [reply]
If the person is already an instance of student, then it behaves like the current mutation. This can be detected by the compiler and translated to `var student = person with {...}`. 

How it would compile? It needs a copy constructor that accepts ascendants. I know that this type of variance is not a language construct, but I don't see any obstacle from IL point of view. For now, I simply added a constructor with Person parameter and it actually works - but one would need plenty of such constructors generated if the hierarchy is targe.

```
public record Person
{
	public string FullName { get; set; }
}

public record Student : Person
{
	public Student(Person master) : base(master) {}
	public string Grade { get; set; }
}

void Main()
{
	var person = new Person { FullName = "XYZ" };
	var s3 = new Student(person) with {Grade = "9"};
}
```

**Reply by HaloFour** [reply]
> This can be detected by the compiler and translated to `var student = person with {...}`.

The compiler doesn't detect this, it relies on whatever type `person` is to do the right thing.

> How it would compile? It needs a copy constructor that accepts ascendants.

That puts the decision of the final type into the hands of the compiler, which is not where it belongs.  What if `person` is a type that inherits from `Student`?  If the compiler emitted `new Student(...)` that would decapitate the instance and you'd lose the additional properties.

**Reply by zorgoz** [reply]
As the syntax involves an _explicit type_ and an _object_, the compiler can certainly rely on these to check the compatibility. This can be done also at runtime with `IsAssignableFtrom` and `IsAssignableTo`. 

```
public static class JustToSeparateThisFromTheRest
{ 
	private static bool IsNotCompatibleWith(this object o1, object o2)
		=> !o1.GetType().IsAssignableFrom(o2.GetType()) && !o1.GetType().IsAssignableTo(o2.GetType());

	internal static void CloneFrom(this object o1, object o2)
	{
		if (o1.IsNotCompatibleWith(o2)) throw new InvalidOperationException();

		var mps = o2.GetType().GetProperties();
		var tps = o1.GetType().GetProperties();

		var mix = mps.Join(tps, m => m.MetadataToken, t => t.MetadataToken, (m, t) => (m, t));
		foreach (var (m, t) in mix)
		{
			t.SetValue(o1, m.GetValue(o2));
		};
	}
}

public record Person
{
	public string FullName { get; set; }
}

public record Student : Person
{
	public Student(object master) => this.CloneFrom(master);

	public Student() {}

	public string Grade { get; set; }
}

public record Senior : Student
{
	public string Role { get; set; }
}

void Main()
{
	var person = new Person { FullName = "XYZ" };

	var senior = new Senior { FullName = "John", Grade = "11", Role = "leader" };

	var student = new Student(person) with {Grade = "9"};
	
	var student2 = new Student(senior) with {Grade = "12"};
	
	person.Dump();
	
	senior.Dump();
	
	student.Dump();
	
	student2.Dump();
}
```

Yes, indeed, if you want to create based on (from) an instance of a derived type, you loose properties. But as the syntax requires you to explicitly state the type you want as result, I don't see any concern here. The syntax again would be: 
`var newInstance = new TargetRecordType from compatibleInstance with {propertyoverrides}`, which would result in generating the "generic constructor" under the hood, like the one I have demonstrated in `Student`.

**Reply by HaloFour** [reply]
A runtime error would be a terrible experience.  If the source is already derived from `Student` why would that ever cause a runtime exception?

**Reply by zorgoz** [reply]
This is only a simulation, a demonstration...

If the source is derived from `Student`, thus the instance is "downgraded" no problem there. The complicated part is the other way around, a.k.a. when "upgrading". There is no language construct I know about, that allows a parameter to accept less specific (ascendant / base) instance than the type specified for a method/constructor parameter (like the delegate contravariance). 
The "compatibility" concept above accepts both directions.
I can only simulate upgrading by using the top level _object_, and this is why I have runtime exception.  However it might still be needed.
